### PR TITLE
chore(auth): Remove `package:oauth2` dependency

### PR DESCRIPTION
**Pull request by dnys1**
The Dart authors are deprecating `package:oauth2` (https://github.com/dart-lang/oauth2/issues/137). This means we can likely not expect to receive further updates. Currently, their `package:http` constraint doesn't allow the latest version (`1.0.0`) and this is the only package we use which brings `http` into our dependency tree.

Rather than force our customers to work around this, this PR has the minimum lift-and-shift required to remove `package:oauth2` and by extension, `package:http`, dependency. It does this by copying over the subset of `package:oauth2` we use and updating it to use `AWSHttpClient` instead.

Accuracy verified through CI and local e2e tests for all platforms.

### CODE REVIEW

File: packages/auth/amplify_auth_cognito_dart/lib/src/oauth/authorization_code_grant.dart
Comment (Equartey) [inline packages/auth/amplify_auth_cognito_dart/lib/src/oauth/authorization_code_grant.dart:1]: Should we also fork the unit tests for these files? I'm fine not taking them if you think our tests are sufficient.

**Review by Equartey — APPROVED** [review]
Good job reducing deps! Just had one non-blocking comment

**Comment by Albert221** [comment]
`oauth2` package version `2.0.2` started to support `http v1.0.0` - https://pub.dev/packages/oauth2/changelog#202

**Comment by dnys1** [comment]
Thanks for the heads up, @Albert221.

Closing in favor of https://github.com/aws-amplify/amplify-flutter/pull/3132. Will reopen when `oauth2` is officially deprecated.

**Pull request closed without merging**